service/settingsservice.php

```diff
@@ -33,16 +34,42 @@ public function __construct(
 	 * @throws \OCP\PreConditionNotMetException
 	 */
 	public function set($settings) {
-		foreach($this->settings as $name => $value) {
-			$this->settings[$name] = isset($settings[$name]) ? $settings[$name] : $value;
+		// remove illegal, empty and default settings
+		foreach($settings as $name => $value) {
+			if(!array_key_exists($name, $this->defaults)
+				|| empty($value)
+				|| $value === $this->defaults[$name]
+			) {
+				unset($settings[$name]);
+			}
 		}
-
-		$this->config->setUserValue($this->uid, 'notes', 'settings', json_encode($this->settings));
+		$this->config->setUserValue($this->uid, 'notes', 'settings', json_encode($settings));
 	}
 
-	public function get($name = null, $default = null) {
+	public function getAll() {
 		$settings = json_decode($this->config->getUserValue($this->uid, 'notes', 'settings'));
-		if(!$settings || !is_object($settings)) $settings = $this->settings;
-		return $name ? (property_exists($settings, $name) ? $settings->{$name} : $default) : $settings;
+		if(is_object($settings)) {
+			// use default for empty settings
+			foreach($this->defaults as $name => $defaultValue) {
+				if(!property_exists($settings, $name) || empty($settings->{$name})) {
+					$settings->{$name} = $defaultValue;
+				}
+			}
+		} else {
+			$settings = (object)$this->defaults;
```

Comment:
i personally would prefer to have a real settings class here and call the constructor with the defaults array. typecasting here makes me have a bad feeling but as i don't know the coding standard in this project and in nextcloud i assume you can ignore this. for projects where i approve the PR i would probably request a change.

in gerneral i would use symfonys option resolver for handling all the settings https://symfony.com/doc/current/components/options_resolver.html

Reply:
We are not using symfony in this app, yet. Personally, I don't want to use another library just for this small thing. But please feel free to provide a PR with your proposed implementation.